src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs

```diff
@@ -0,0 +1,135 @@
+
+            private readonly HttpContext _httpContext;
+            private IEnumerable<KeyValuePair<string, object>> _cachedGetValues;
+            private string _cachedToString;
+
+            public HostingRequestStarting(HttpContext httpContext)
+            {
+                _httpContext = httpContext;
+            }
+
+            public override string ToString() => _cachedToString ?? Interlocked.CompareExchange(
+                ref _cachedToString,
+                $"Request starting {_httpContext.Request.Protocol} {_httpContext.Request.Method} {_httpContext.Request.Scheme}://{_httpContext.Request.Host}{_httpContext.Request.PathBase}{_httpContext.Request.Path}{_httpContext.Request.QueryString} {_httpContext.Request.ContentType} {_httpContext.Request.ContentLength}",
+                null);
+
+            public IEnumerable<KeyValuePair<string, object>> GetValues() => _cachedGetValues ?? Interlocked.CompareExchange(
+                ref _cachedGetValues,
+                new[]
+                {
+                    new KeyValuePair<string, object>("EventName", "RequestStarting"),
+                    new KeyValuePair<string, object>("Protocol", _httpContext.Request.Protocol),
+                    new KeyValuePair<string, object>("Method", _httpContext.Request.Method),
+                    new KeyValuePair<string, object>("ContentType", _httpContext.Request.ContentType),
+                    new KeyValuePair<string, object>("ContentLength", _httpContext.Request.ContentLength),
+                    new KeyValuePair<string, object>("Scheme", _httpContext.Request.Scheme.ToString()),
+                    new KeyValuePair<string, object>("Host", _httpContext.Request.Host.ToString()),
+                    new KeyValuePair<string, object>("PathBase", _httpContext.Request.PathBase.ToString()),
+                    new KeyValuePair<string, object>("Path", _httpContext.Request.Path.ToString()),
+                    new KeyValuePair<string, object>("QueryString", _httpContext.Request.QueryString.ToString()),
+                },
+                null);
+        }
+
+        private class HostingRequestFinished
+        {
+            internal static readonly Func<object, Exception, string> Callback = (state, exception) => ((HostingRequestFinished)state).ToString();
+
+            private readonly HttpContext _httpContext;
+            private IEnumerable<KeyValuePair<string, object>> _cachedGetValues;
+            private string _cachedToString;
+
+            public HostingRequestFinished(HttpContext httpContext)
+            {
+                _httpContext = httpContext;
+            }
+
+            public override string ToString() => _cachedToString ?? Interlocked.CompareExchange(
+                ref _cachedToString,
+                $"Request finished {_httpContext.Response.StatusCode} {_httpContext.Response.ContentType} {_httpContext.Response.Body.Length}",
```

Comment:
Per discussion with lou, going to take this out and follow up later on adding the length here.